### ISSUE TITLE
Need to be able to learn AxisFunctions

### ISSUE THREAD

**Issue by gabbard** [isi-vista/adam#712]

**Comment by gabbard** [comment]
@LichMaster98  and @denizbeser : can you each spend ~30 minutes thinking about how to do this and write up your conclusions? (which could be "I couldn't figure out a way to do it" :-) )

**Comment by denizbeser** [comment]
If I understand correctly, axis function become necessary once we have geometry involved in learning. So, it could be good to have axes functions (or some of them) ignored when we are learning objects, and start "perceiving" them when we encounter situations where they matter. We could go with a layered/iterative approach such that the model starts paying attention to them _once it understand _not_ paying attention to them is insufficient_. This is some sort of "specialization" in perceptive sense, just like how we get better and better at hearing harmony more we listen to music.

**Comment by gabbard** [comment]
@denizbeser : I think you are correct that the learner can reasonable "just know" that axis functions are irrelevant for objects and only matter for relations and actions.  We still need to move towards a more concrete strategy for learning then, though. :-)

**Comment by denizbeser** [comment]
@gabbard aren't they available in the perception graphs already? What stops pursuit from learning them?

**Comment by gabbard** [comment]
@denizbeser : they are not, at least not directly. An `AxisFunction` computes something like "whichever axis of an object faces a the learner"). Or at least we need to enumerate the `AxisFunctions` we have and determine which are straightforwardly learnable from the perceptions.

**Comment by LichMaster98** [comment]
@gabbard & @denizbeser :
Our current Axis Functions are:
- PrimaryAxisOfObject
- HorizontalAxisOfObject
- FacingAddresseeAxis
- GRAVITATIONAL_AXIS_FUNCTION

Learning `PrimaryAxisOfObject` and `HorizontalAxisOfObject` seem straightforward as these Axes are always easily identifiable by the learned objects. For learning this axis relationship however between multiple objects in an action I'm unsure if we currently provide a way of saying two axis are parallel & overlapping (similar?) so that either could be used as the reference object to get the same axis result. (Imagine two boxes sitting next to each other, I could say that `box_1` is on the negative direction of the horizontal axis using the reference object of `box_2` or I could say that `box_1` is on the positive direction of the horizontal axis using the reference object of `box_1`. While these axis are distinct by definition in practice they convey the same line. I'm unsure if we ever have a situation like this in our curriculum.

Learning the `GRAVITATIONAL_AXIS_FUNCTION` similarly seems straightforward as the axis itself for that function is concrete. I'm going to assume the axis of Gravity is unchanging for our world. :) 

To learn `FacingAddresseeAxis` I believe we'll need to address #573 (Additionally, I'm unsure if https://github.com/isi-vista/adam/blob/master/adam/situation/templates/phase1_templates.py#L708-#L714 is a problem to consider for this issue) 

The issue I see here is that the axis connecting the addressee and the two relevant objects in this relationship aren't straightforward. We're stuck with a similar problem as the axis similarity issue seen in `HorizontalAxis` & `PrimaryAxis`. I didn't come up to a solution here but I believe we'll need to extend the amount of information available in the `axis_info` to make learning this easier. Perhaps this: #680 is causing issues resolving the axis relationships during an action? I'm unsure how but at least considering it as a potential problem in learning AxisFunctions seems wise.

**Comment by gabbard** [comment]
@LichMaster98 : `PrimaryAxisOfObject` can be read directly off the perception graph due to the handy `primary-axis` label.  But how can we easily identify the horizontal axis from the current representation?  (Where do we use `HorizontalAxisOfObject`? Maybe we can eliminate the need for it altogether?)

I agree the gravitational axis is easy, and in fact I think we already succeed in learning it just fine (since it is the `primary-axis` of the `ground` object, IIRC).

For #573, would is be sufficient to add edges between the axes and the learning encoding the facing information? The other problem you note is not a problem directly but might lead to learning a spurious correlation between object horizontal axes and facings.

> The issue I see here is that the axis connecting the addressee and the two relevant objects in this relationship aren't straightforward. We're stuck with a similar problem as the axis similarity issue seen in HorizontalAxis & PrimaryAxis. I didn't come up to a solution here but I believe we'll need to extend the amount of information available in the axis_info to make learning this easier.

Can you clarify what "this relationship" refers to?

re: #680 - I think it is only a problem when objects changes locations between frames in a way which alters their axis relationships.

**Comment by LichMaster98** [comment]
@gabbard We use `HorizontalAxisOfObject` for identifying the perception of "beside" 

For "this relationship" between the `FacingAddresseeAxis` I was trying to identify that the axis that gives the information for "a ball in front of a box" doesn't refer to either of the two objects in the language, which is a difference from most of our other prepositions. Currently I think our hypothesis space for the relationship then wouldn't collect the reference object of the axis of importance.

**Comment by gabbard** [comment]
Is the next step here to ensure that during perceptual generate we relate every object to the learner via some region relationship using that object as the reference object and the axis which faces the learner as the axis of the relationship?

**Comment by LichMaster98** [comment]
@gabbard Possibly, though that may continue our trend of spurious learning of non-essential relationships to the learner for all scenes and objects.

**Comment by gabbard** [comment]
@LichMaster98 : True. I think we should go ahead, keeping in mind it will boost the priority of https://github.com/isi-vista/adam/issues/696
